syntax = "proto3";

option java_package = "rescuecore2.messages.protobuf";
option java_outer_classname = "ControlMessageProto";

message MessageProto {
   string urn = 1;
   map<string,MessageComponentProto> components=2;
}
message MessageListProto{
	repeated MessageProto commands = 1;	
}

message MessageComponentProto {
   oneof component {
		   ChangeSetProto changeSet=1;
		   MessageListProto commandList=2;
		   ConfigProto config=3;
		   EntityProto entity=4;
		   int32 entityID=5;
		   IntListProto entityIDList=6;
		   EntityListProto entityList=7;
		   FloatListProto floatList=8;
		   int32 intValue=9;
		   IntListProto intList=10;
		   bytes rawData=11;
		   string stringValue=12;
		   StrListProto stringList=13;
  }
}


message StrListProto {
  repeated string values = 1;
}

message IntListProto {
  repeated int32 values = 1;
}

message FloatListProto {
	  repeated float values = 1;
}

message IntMatrixProto {
  repeated IntListProto values = 1;
}

message ValueProto {
  enum Name {
	  Value=0;
	  X=1;
	  Y=2;
	  EntityRef=3;
	  EntityRefList=4;
  }
  Name name = 1;
  oneof value {
    int32 valueInt = 2;
    bool valueBool = 3;
    double valueDouble = 4;
    bytes listByte = 5;
    IntListProto listInt = 6;
    IntMatrixProto matrixInt = 7;
    EdgeListProto listEdge = 8; 
  }
}

message PropertyProto {
  string urn = 1;
  bool defined = 2;
  repeated ValueProto fields = 3;
}

message EntityProto {
  string urn = 1;
  int32 entityID = 2;
  map<string,PropertyProto> properties = 3;
}

message EntityListProto {
	repeated EntityProto entities = 1;	
}
message ConfigProto {
  map<string, string> data = 1;
}

message EdgeListProto {
	repeated EdgeProto edges=1;
};
message EdgeProto {
	int32 startX = 1;
    int32 startY = 2;
    int32 endX = 3;
    int32 endY = 4;
    int32 neighbour = 5;
};

message ChangeSetProto {
	message EntityChangeProto {
		int32 entityID=1;
		string urn=2;
		repeated PropertyProto properties = 3;
	}
  repeated EntityChangeProto changes = 1;
  repeated int32 deletes = 2; /*entity ids*/
}



message CommandLogProto {
  int32 time = 1;
  repeated MessageProto commands = 2;
}

message InitialConditionsLogProto {
  repeated EntityProto entities = 1;
}

message PerceptionLogProto {
  int32 time = 1;
  int32 entityID = 2;
  ChangeSetProto visible = 3;
  repeated MessageProto communications = 4;
}

message StartLogProto {
}

message EndLogProto {
}

message ConfigLogProto {
  ConfigProto config = 1;
}

message UpdatesLogProto {
  int32 time = 1;
  ChangeSetProto changes = 2;
}